mqtt/mqtt-broker/src/edgehub_auth.rs

```diff
@@ -0,0 +1,370 @@
+use async_trait::async_trait;
+use base64::encode;
+use failure::ResultExt;
+use serde::{Deserialize, Serialize};
+use std::fmt;
+
+use crate::auth::{AuthId, Authenticator, Credentials};
+use crate::{Error, ErrorKind};
+
+pub struct EdgeHubAuthenticator {
+    url: String,
+}
+
+const API_VERSION: &str = "2020-04-20";
+
+#[async_trait]
+impl Authenticator for EdgeHubAuthenticator {
+    async fn authenticate(
+        &self,
+        username: Option<String>,
+        credentials: Credentials,
+    ) -> Result<Option<AuthId>, Error> {
+        self.internal_authenticate((username, credentials).into())
+            .await
+            .context(ErrorKind::AuthExecution)?
+            .into_auth_id()
+    }
+}
+
+impl EdgeHubAuthenticator {
+    async fn internal_authenticate(
+        &self,
+        request: EdgeHubAuthRequest,
+    ) -> Result<EdgeHubAuthResponse, Error> {
+        let response = reqwest::Client::new()
+            .post(&self.url)
+            .json(&request)
+            .send()
+            .await
+            .context(ErrorKind::AuthSendRequest)?;
+
+        if response.status() != reqwest::StatusCode::OK {
+            return Err(ErrorKind::AuthProcessResponse.into());
+        }
+
+        let response = response
+            .json::<EdgeHubAuthResponse>()
+            .await
+            .context(ErrorKind::AuthProcessResponse)?;
+
+        if !response.version.eq(API_VERSION) {
+            return Err(ErrorKind::AuthApiVersion(response.version).into());
+        }
+
+        Ok(response)
+    }
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "snake_case")]
+pub struct EdgeHubAuthRequest {
+    version: String,
+    username: Option<String>,
+    password: Option<String>,
+    certificate: Option<String>,
+    certificate_chain: Option<Vec<String>>,
+}
+
+impl From<(Option<String>, Credentials)> for EdgeHubAuthRequest {
+    fn from(item: (Option<String>, Credentials)) -> Self {
+        let (password, certificate) = match item.1 {
+            Credentials::Password(p) => (p, None),
```

Comment:
as far as I can see, this line will always match. Maybe you wanted `Credentials::Password(**Some(p)**) => (p, None),`